SECEdgar/extractor/EDGARExtractor.py

```diff
@@ -0,0 +1,227 @@
+import re
+import json
+import os.path
+import logging
+import uu
+
+
+class EDGARExtractor():
+
+    # Compile regular expressions
+    # Note that re.search() does not have a starting position argument
+    # but the search() method of a Pattern object
+    # (a compiled regex) has a pos argument
+    re_doc = re.compile("<DOCUMENT>(.*?)</DOCUMENT>", flags=re.DOTALL)
+    re_sec_header = re.compile("<SEC-HEADER>.*?\n(.*?)</SEC-HEADER>",
+                               flags=re.DOTALL)
+    re_sec_doc = re.compile("<SEC-DOCUMENT>(.*?)</SEC-DOCUMENT>",
+                            flags=re.DOTALL)
+    re_text = re.compile("<TEXT>(.*?)</TEXT>", flags=re.DOTALL)
+
+    def __init__(self, in_file, out_dir=None, metadata_debug=False,
+                  create_subdir=True, rm_infile=False):
+        """
+        Extract the embedded documents and
+        their respective metadata from an EDGAR txt file
+        """
+        self.metadata_debug = metadata_debug
+        if out_dir is None:
+            out_dir = os.path.dirname(os.path.abspath(in_file))
+        self.procTxtFile(in_file, out_dir,
+                         create_subdir=create_subdir,
+                         rm_infile=rm_infile)
+
+    def procTxtFile(self, in_file, out_dir, create_subdir=True, rm_infile=False):
```

Comment:
I feel like the `EdgarExtractor` class is doing more than what I understand its responsibilities are.
1. It is trying to do regex operation
2. It is trying to save file
3. The matching and searching of the content.
4. And other small things. 

IMO, we can extract all these things. Extractor responsibility will get the text file and return the expected context.

```
class FilingExtractor():
      ___init__(regex: Regex()): Class will take dependency for parsers or regexes
      
      def extract(content): // this should return the final content
```

Reply:
I am not understanding your point fully
1. Regex operations are needed to extract the content of the .txt file. The regex is getting the content between some tags and putting it in a different file
2. Ok, we could pass the content in memory without saving. But I think it doesn't harm to have a 
bool that if set to True will save the files? The "Filings" class has a save() method too
3. I don't understanding this. Isn't it analogous to point 1?

`Extractor responsibility will get the text file and return the expected context.`

It can return a list with the contents of each file as its elements, and another parameter with the metadata (like the filenames of every file in the first list). But this would cover point 2 of your notes. I don't get point 1,3 and 4. You want to extract the content of the .txt files without using regex?

Reply:
I think that maybe `proc_text_file` can return an object of another class, called something like `ExtractedFiling` maybe. This other class can contain the metadata and the files extracted, and a method to save the metadata and/or the files extracted to disk. What do you think about that?

Reply:
I want to extract the content using regex. We have two one class `EdgarExtractor` here which is doing everything.

let break it into two or three as per requirement.

1. The `FilingExtractor` will have a dependency on the RegexOperation class as mentioned in the above comment. This way we can easily test what regex operations are doing. 
2. We can also have a separate class related to file operations. Maybe we can name it as `Writer`. It will handle writing the content on file at the correct location.

Now FilingExtractor will look like this

```
class FilingExtractor(object):

   def __init__(self, regexOperation, writer, filing_data):
         self. regexOperation = regexOperation
         self.writer = writer
         self.filing_data = filing_data

    def extract(self):
         content = self.regexOperation(filing_data)
         return content
    
     def writeToDisk(self, path):
          self.writer(content, path)
 

```

This way we break the responsibilities and one class does one thing. It will make our code much easier to test. Hope this clarifies.